File: frontend/src/services/categoryService.js
import API from './api';

export const categoryService = {
  getCategories: async () => {
    const res = await API.get('/categories');
    return res.data;
  },

  getCategoryById: async (id) => {
    const res = await API.get(`/categories/${id}`);
    return res.data;
  },

  createCategory: async (data) => {
    const res = await API.post('/categories', data);
    return res.data;
  },

  updateCategory: async (id, data) => {
    const res = await API.put(`/categories/${id}`, data);
    return res.data;
  },

  deleteCategory: async (id) => {
    const res = await API.delete(`/categories/${id}`);
    return res.data;
  }
};

File: backend/src/main/java/com/shopsphere/service/CategoryService.java
package com.shopsphere.service;

import com.shopsphere.dto.CategoryRequest;
import com.shopsphere.exception.BadRequestException;
import com.shopsphere.exception.ResourceNotFoundException;
import com.shopsphere.model.Category;
import com.shopsphere.repository.CategoryRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class CategoryService {

    @Autowired
    private CategoryRepository categoryRepository;

    public List<Category> getAllCategories() {
        return categoryRepository.findAll();
    }

    public Category getCategoryById(String id) {
        return categoryRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Category not found with id: " + id));
    }

    public Category createCategory(CategoryRequest request) {
        if (categoryRepository.existsByNameIgnoreCase(request.getName())) {
            throw new BadRequestException("Category already exists with name: " + request.getName());
        }

        Category category = Category.builder()
                .name(request.getName())
                .description(request.getDescription())
                .imageUrl(request.getImageUrl())
                .build();

        return categoryRepository.save(category);
    }

    public Category updateCategory(String id, CategoryRequest request) {
        Category category = getCategoryById(id);

        category.setName(request.getName());
        category.setDescription(request.getDescription());
        if (request.getImageUrl() != null) {
            category.setImageUrl(request.getImageUrl());
        }

        return categoryRepository.save(category);
    }

    public void deleteCategory(String id) {
        Category category = getCategoryById(id);
        categoryRepository.delete(category);
    }
}
